Common/Source/Dialogs/dlgFlarmIGCDownload.cpp

```diff
@@ -0,0 +1,897 @@
+    	 if(wf) wf->SetTimerNotify(0, NULL);    // reset Timer
+		 switch (DownloadError)
+		 {
+		   case REC_NO_ERROR     : _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s\n%s"),  szIGCStrings[IGC_Index] ,MsgToken(2406)); break; // 	_@M2406_ "IGC File download complete"
+		   case REC_TIMEOUT_ERROR: _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s"), MsgToken(2407)); break;  // _@M2407_ "Error: receive timeout
+		   case REC_CRC_ERROR    : _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s"), MsgToken(2408)); break;  // _@M2408_ "Error: CRC checksum fail"
+		   case REC_ABORTED      : _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s"), MsgToken(2415)); break;  // _@M2415_ "IGC Download aborted!"
+		   case FILENAME_ERROR   : _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s"), MsgToken(2409)); break;  // _@M2409_ "Error: invalid filename"
+		   case FILE_OPEN_ERROR  : _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s"), MsgToken(2410)); break;  // _@M2410_ "Error: can't open file"
+		   case IGC_RECEIVE_ERROR: _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s"), MsgToken(2411)); break;  // _@M2411_ "Error: Block invalid"
+		   case REC_NO_DEVICE    : _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s"), MsgToken(2401)); break;  // _@M2401_ "No Device found"
+
+
+		   default               : _sntprintf(Tmp, STATUS_TXT_LEN, _T("%s"), MsgToken(2412)); break;  // _@M2412_ "Error: unknown"
+
+		 }
+		 if (MessageBoxX( Tmp, MsgToken(2398), mbOk) == IdYes)  // _@M2406_ "Error: communication timeout"Reboot"
+		 {
+		 }
+		 DownloadError = REC_NO_ERROR;
+	   }
+       else
+       {
+#ifdef PRPGRESS_DLG
+         IGCProgressDialogText(szStatusText) ;  // update progress dialog text
+#endif
+       }
+    }
+  }
+  return true;
+}
+
+
+static CallBackTableEntry_t IGCCallBackTable[] = {
+    OnPaintCallbackEntry(OnMultiSelectListPaintListItem),
+    OnListCallbackEntry(OnMultiSelectListListInfo),
+    ClickNotifyCallbackEntry(OnCloseClicked),
+    ClickNotifyCallbackEntry(OnUpClicked),
+    ClickNotifyCallbackEntry(OnEnterClicked),
+    ClickNotifyCallbackEntry(OnDownClicked),
+
+
+	EndCallBackEntry()
+};
+
+
+
+ListElement* dlgIGCSelectListShowModal( DeviceDescriptor_t *d) {
+
+MapWindow::SuspendDrawingThread();
+StartupStore(TEXT(".... StartIGCReadThread%s"),NEWLINE);
+StartIGCReadThread() ;
+
+for (int i = 0; i < MAX_IGCFILES; i++)
+{
+  szIGCStrings[i] = NULL;
+  szIGCSubStrings[i] = NULL;
+}
+
+szIGCStrings[0]    = new TCHAR[LST_STRG_LEN];
+szIGCSubStrings[0] = new TCHAR[LST_STRG_LEN];
+IGCFilename        = new TCHAR[MAX_PATH];
+if(IGCFilename == NULL) return NULL;
+
+	IGC_Index = -1;
+	iNoListLine =-1;
+	uint16_t RecSequence;
+	uint8_t RecCommand;
+	uint8_t pBlock[100];
+	uint16_t blocksize;
+	TCHAR  TempString[255];
+
+    LockComm();
+
+    /*************************************************/
+    ThreadState =  OPEN_BIN_STATE;
+    /*************************************************/
+    wf = dlgLoadFromXML(IGCCallBackTable, ScreenLandscape ? IDR_XML_MULTISELECTLIST_L : IDR_XML_MULTISELECTLIST_P);
+
+    if (!wf) return NULL;
```

Comment:
- memory leak 
- missing unlock Mutex
- missing Resume Drawing thread

this is a good exemple of one of common mistake easily avoided if you use the simple RAII pattern
https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization